src/Npgsql/Internal/Resolvers/UnsupportedTypeResolver.cs

```diff
@@ -13,8 +14,18 @@ sealed class UnsupportedTypeInfoResolver<TBuilder> : IPgTypeInfoResolver
         if (typeof(IEnumerable<>).IsAssignableFrom(type) && !typeof(IList).IsAssignableFrom(type) && type != typeof(string))
             throw new NotSupportedException("Writing is not supported for IEnumerable parameters, use an array or List instead.");
 
-        if (type != typeof(object) && DataTypeNames.Record == dataTypeName)
-            throw new NotSupportedException(string.Format(NpgsqlStrings.RecordsNotEnabled, "EnableRecords", typeof(TBuilder).Name));
+        if (type != typeof(object) && dataTypeName == DataTypeNames.Record)
+        {
+            throw new NotSupportedException(
+                string.Format(NpgsqlStrings.RecordsNotEnabled, nameof(NpgsqlSlimDataSourceBuilder.EnableRecords), typeof(TBuilder).Name));
+        }
+
+        if (type is { IsGenericType: true } && type.GetGenericTypeDefinition() == typeof(NpgsqlRange<>)
+            || dataTypeName.HasValue && options.TypeCatalog.GetPostgresTypeByName(dataTypeName.Value) is PostgresRangeType)
```

Comment:
```suggestion
            || (typeof(T) != typeof(object) && dataTypeName.HasValue && options.TypeCatalog.GetPostgresTypeByName(dataTypeName.Value) is PostgresRangeType))
```

This is to allow the objectordefault probing to succeed, as we start looking for an object converter first before calling into the resolvers again with type = null.

Might want to add a test for that

Reply:
Alternatively we can add a try catch to CreateInfo in the cache, would mean we swallow potentially legitimate errors during that fallback enabled lookup.

Reply:
Take a look, I've pushed something that IMO is not tenable but I'm not sure how we improve this without throwing a lot of false positives.

Reply:
Looks good, thanks.

BTW this raises the cross-cutting problem with NodaTime ranges, i.e. does adding NodaTime automatically add ranges. We could in theory do something like for arrays, where only doing both AddNodaTime and EnableRanges gives you these, but it's probably more trouble than its worth (we're only talking about 3 ranges for the NodaTime case, we can just throw them in).

(if we do want to handle it like for arrays, that would raise the question of raising the right "unsupported ranges" error for NodaTime arrays, given that you're hard-coding a list of specific CLR types into RangeTypeInfoResolver; if we went down that road we could consider throwing the range error message for any range instead).

But I think the current situation is fine.